syntax="proto3";

message BallotNum {
   int32 num = 1;
   int32 pid = 2;
   int32 depth = 3;
}

message Init {
   int32 src = 1;
}

//Type: 1
message Prepare {
   int32 type = 1;
   BallotNum ballot = 2;
}

message Transaction {
   int32 src = 1;
   int32 rcvr = 2;
   int32 amt = 3;
}

message Block {
   repeated Transaction trans = 2;
   string nonce = 3;
   string hash = 4; 
}

//Type: 3
message Promise {
   int32 type = 1;
   BallotNum ballot = 2;
   BallotNum acceptNum = 3;
   Block acceptVal = 4;
}


//Type: 4
message Accept {
   int32 type = 1;
   BallotNum ballot = 2;
   Block myVal = 3;
}

//Type: 5
message Accepted {
   int32 type = 1;
   BallotNum ballot = 2;
   Block acceptVal = 3;
}

//Type: 6
message Decide {
   int32 type = 1;
   int32 src = 4;
   BallotNum ballot = 2;
   Block val = 3;
}  

//Type: 7
message Recover {
   int32 type = 1;
}

//Type: 8
message RepRecover {
   int32 type = 1;
   repeated Block insi = 2;
}
